ICSharpCode.Decompiler/IL/Transforms/NullPropagationTransform.cs

```diff
@@ -319,6 +319,218 @@ public void Run(Block block, int pos, StatementTransformContext context)
+				// if the ancestors do not yet have a NullableRewrap instruction
+				// insert it at the 'rewrapPoint'.
+				if (needsRewrap) {
+					siblings[index] = new NullableRewrap(rewrapPoint);
+				}
+				// if the endBlock is only reachable through the current block,
+				// combine both blocks.
+				if (endBlock?.IncomingEdgeCount == 1) {
+					block.Instructions.AddRange(endBlock.Instructions);
+					block.Instructions.RemoveAt(pos + 1);
+					endBlock.Remove();
+				}
+			}
+		}
+
+		// stloc valueTemporary(valueExpression)
+		// stloc defaultTemporary(default.value type)
+		// if (logic.not(comp.o(box `0(ldloc defaultTemporary) != ldnull))) Block fallbackBlock {
+		// 	stloc defaultTemporary(ldobj type(ldloc valueTemporary))
+		// 	stloc valueTemporary(ldloca defaultTemporary)
+		// 	if (comp.o(ldloc defaultTemporary == ldnull)) Block fallbackBlock2 {
+		// 		stloc resultTemporary(ldnull)
+		// 		br endBlock
+		// 	}
+		// }
+		// stloc resultTemporary(constrained[type].call_instruction(ldloc valueTemporary, ...))
+		// br endBlock
+		// =>
+		// stloc resultTemporary(nullable.rewrap(constrained[type].call_instruction(nullable.unwrap(valueExpression), ...)))
+		//
+		// -or-
+		//
+		// stloc valueTemporary(valueExpression)
+		// stloc defaultTemporary(default.value type)
+		// if (logic.not(comp.o(box `0(ldloc defaultTemporary) != ldnull))) Block fallbackBlock {
+		// 	stloc defaultTemporary(ldobj type(ldloc valueTemporary))
+		// 	stloc valueTemporary(ldloca defaultTemporary)
+		// 	if (comp.o(ldloc defaultTemporary == ldnull)) Block fallbackBlock2 {
+		// 		leave(ldnull)
+		// 	}
+		// }
+		// leave (constrained[type].call_instruction(ldloc valueTemporary, ...))
+		// =>
+		// leave (nullable.rewrap(constrained[type].call_instruction(nullable.unwrap(valueExpression), ...)))
+		private bool TransformNullPropagationOnUnconstrainedGenericExpression(Block block, int pos,
+			out ILInstruction target, out ILInstruction value, out ILInstruction rewrapPoint, out Block endBlock)
+		{
+			target = null;
+			value = null;
+			rewrapPoint = null;
+			endBlock = null;
+			if (pos + 3 >= block.Instructions.Count)
+				return false;
+			// stloc valueTemporary(valueExpression)
+			if (!(block.Instructions[pos].MatchStLoc(out var valueTemporary, out value)))
+				return false;
+			if (!(valueTemporary.Kind == VariableKind.StackSlot && valueTemporary.LoadCount == 2 && valueTemporary.StoreCount == 2))
+				return false;
+			// stloc defaultTemporary(default.value type)
+			if (!(block.Instructions[pos + 1].MatchStLoc(out var defaultTemporary, out var defaultExpression) && defaultExpression.MatchDefaultValue(out var type)))
+				return false;
+			// Some variables are reused by the compiler therefore we cannot check for absolute numbers. See for example ValueTuple`8.ToString
+			// LoadCount must be == StoreCount and LoadCount must be 2x AddressCount.
+			// Note: if successful, this transform removes exactly 2 store, 2 load and 1 addressof instruction.
```

Comment:
But this only works if indeed all uses of the variable follow the expected pattern.
Otherwise it could be possible that deleting a store changes the semantics of one of the other loads.

I don't think we can support this variable re-use case without introducing bugs here.
If we're going to support it at all, it's going to have to involve variable splitting.

Reply:
So, the suggestion is to go back to checking for exact load/store/address counts?

Reply:
Yes, the transform should only apply in cases where we can argue that it's correct. We can't do that just based on some multiple of the expected accesses without actually checking those other accesses.

Reply:
I just tested `ValueTuple`8.ToString` with 060830dd6460afbd21a4f6731ee2995bc8f20fc6 applied to this branch. Unfortunately, this does not solve the problem.

Would it be possible for you to take a look at this?